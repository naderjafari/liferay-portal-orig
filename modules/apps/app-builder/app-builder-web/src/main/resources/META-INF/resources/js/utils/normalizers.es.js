/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {DataConverter} from 'data-engine-taglib';

import {getValidName} from '../utils/utils.es';

export const normalizeNames = ({
	allowEmptyKeys = true,
	defaultName = '',
	localizableValue,
}) => {
	const name = {};

	Object.keys(localizableValue).forEach((languageId) => {
		const value = localizableValue[languageId];
		const normalizedValue = getValidName(defaultName, value)?.trim();

		if (!allowEmptyKeys && !normalizedValue) {
			return;
		}

		name[languageId] = normalizedValue;
	});

	return name;
};

export const normalizeDataDefinition = (dataDefinition) => ({
	...dataDefinition,
	dataDefinitionFields: dataDefinition.dataDefinitionFields.map(
		(dataDefinitionField) => ({
			...dataDefinitionField,

			// Actually showLabel property will be always true
			// because the same property can be controlled by dataLayoutFields

			showLabel: true,
		})
	),
});

export const normalizeDataLayout = ({
	dataDefinition,
	dataLayout,
	dataLayoutBuilder,
	defaultLanguageId,
	editingLanguageId,
}) => {
	const {dataDefinitionFields = []} = dataDefinition;
	const dataLayoutFields = {...dataLayout.dataLayoutFields};

	Object.keys(dataLayoutFields).forEach((field) => {
		if (!dataDefinitionFields.find((item) => item.name === field)) {
			delete dataLayoutFields[field];
		}
	});

	dataDefinitionFields.forEach((dataDefinitionField) => {
		const {
			appContext: [{editingLanguageId}],
			fieldTypes,
		} = dataLayoutBuilder.props;
		const fieldProperties = DataConverter.getDDMSettingsContextWithVisualProperties(
			{
				dataDefinitionField,
				editingLanguageId,
				fieldTypes,
			}
		);

		// Ignore this visual properties because it is treated differently

		delete fieldProperties['label'];
		delete fieldProperties['required'];

		dataLayoutFields[dataDefinitionField.name] = {
			...dataLayoutFields[dataDefinitionField.name],
			...fieldProperties,
		};
	});

	const name = normalizeNames({
		defaultName: Liferay.Language.get('untitled-form-view'),
		localizableValue: dataLayout.name,
	});

	if (!name[defaultLanguageId]) {
		name[defaultLanguageId] = name[editingLanguageId];
	}

	return {
		...dataLayout,
		dataLayoutFields,
		name,
	};
};
