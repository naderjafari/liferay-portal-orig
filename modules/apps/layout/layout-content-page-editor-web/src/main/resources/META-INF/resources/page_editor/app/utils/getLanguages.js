/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {config} from '../../app/config/index';

export default function getLanguages(
	availableLanguages,
	availableSegmentsExperiences,
	segmentsExperienceId
) {
	if (
		!availableSegmentsExperiences ||
		!segmentsExperienceId ||
		!availableSegmentsExperiences[segmentsExperienceId] ||
		segmentsExperienceId === config.defaultSegmentsExperienceId
	) {
		return availableLanguages;
	}

	const languages =
		availableSegmentsExperiences[segmentsExperienceId].languageIds;

	return languages.reduce((acc, languageId) => {
		if (availableLanguages[languageId]) {
			return {
				...acc,
				[languageId]: {...availableLanguages[languageId]},
			};
		}

		return {...acc};
	}, {});
}
