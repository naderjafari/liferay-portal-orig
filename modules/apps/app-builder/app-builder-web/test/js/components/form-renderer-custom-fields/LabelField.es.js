/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

import {cleanup, fireEvent, render} from '@testing-library/react';
import React from 'react';

import LabelField from '../../../../src/main/resources/META-INF/resources/js/components/form-renderer-custom-fields/LabelField.es';
import {getFormattedState} from '../../../../src/main/resources/META-INF/resources/js/components/form-renderer-custom-fields/shared/utils.es';
import {FORM_VIEW} from '../../constants.es';

const DEFAULT_STATE = {
	editingLanguageId: 'en_US',
	focusedCustomObjectField: {},
	focusedField: {
		fieldName: 'TextName',
	},
};

const DATA_DEFINITION_FIELD = (label) => ({
	label: {
		en_US: label,
	},
	name: `Text${label}`,
});

const STRUCTURE_LEVEL_STATE = {
	...DEFAULT_STATE,
	dataDefinition: {
		dataDefinitionFields: [
			{
				...DATA_DEFINITION_FIELD('Name'),
				customProperties: {
					labelAtStructureLevel: true,
				},
			},
			{
				...DATA_DEFINITION_FIELD('Email'),
				customProperties: {
					labelAtStructureLevel: true,
				},
			},
		],
	},
	dataLayout: {
		dataLayoutFields: {},
	},
};

const VIEW_LEVEL_STATE = {
	...DEFAULT_STATE,
	dataDefinition: {
		dataDefinitionFields: [
			{
				...DATA_DEFINITION_FIELD('Name'),
				customProperties: {
					labelAtStructureLevel: false,
				},
			},
			{
				...DATA_DEFINITION_FIELD('Email'),
				customProperties: {
					labelAtStructureLevel: false,
				},
			},
		],
	},
	dataLayout: {
		dataLayoutFields: {
			TextEmail: {
				label: {
					en_US: 'Login',
				},
			},
			TextName: {
				label: {
					en_US: 'Username',
				},
			},
		},
	},
};

const LabelFieldWrapper = ({
	state = STRUCTURE_LEVEL_STATE,
	dataLayoutBuilder = FORM_VIEW.getDataLayoutBuilderProps(),
}) => (
	<LabelField
		dataLayoutBuilder={dataLayoutBuilder}
		dispatch={jest.fn()}
		field={{
			label: 'Label',
			name: 'LabelName',
			placeholder: 'Enter a field label.',
			tooltip:
				'Enter a descriptive field label that guides users to enter the information you want.',
		}}
		state={getFormattedState(state)}
	/>
);

describe('LabelField', () => {
	beforeEach(() => {
		jest.useFakeTimers();
	});

	afterEach(() => {
		cleanup();

		jest.clearAllTimers();
		jest.restoreAllMocks();
	});

	afterAll(() => {
		jest.useRealTimers();
	});

	it('renders with structure-level', () => {
		const {asFragment} = render(<LabelFieldWrapper />);

		expect(asFragment()).toMatchSnapshot();
	});

	it('renders with view-level', () => {
		const {asFragment} = render(
			<LabelFieldWrapper state={VIEW_LEVEL_STATE} />
		);
		expect(asFragment()).toMatchSnapshot();
	});

	it('renders with structure-level passing the focusedCustomObjectField', () => {
		const {container} = render(
			<LabelFieldWrapper
				state={{
					...STRUCTURE_LEVEL_STATE,
					focusedCustomObjectField: {
						name: 'TextName',
					},
					focusedField: {},
				}}
			/>
		);

		const input = container.querySelector('.ddm-field-text');

		expect(input.value).toBe('Name');
	});

	it('display structure-level label and do not render the object-field-label message', () => {
		const {container, queryByText} = render(<LabelFieldWrapper />);

		const input = container.querySelector('.ddm-field-text');

		expect(input.value).toEqual('Name');
		expect(queryByText('object-field-label-x')).toBeFalsy();
	});

	it('display view-level label and render the message object-field-label', () => {
		const {container, queryByText} = render(
			<LabelFieldWrapper state={VIEW_LEVEL_STATE} />
		);

		const input = container.querySelector('.ddm-field-text');

		expect(input.value).toEqual('Username');
		expect(queryByText('object-field-label-x')).toBeTruthy();
	});

	it('renders the popover with structure level option selected', () => {
		const {container} = render(<LabelFieldWrapper />);

		const dropdownButton = container.querySelector(
			'.form-renderer-label-field__button'
		);
		const popover = container.querySelector(
			'.form-renderer-label-field__popover'
		);
		const options = document.querySelectorAll(
			'.form-renderer-label-field input.custom-control-input'
		);

		expect(popover.classList.contains('show')).toBeFalsy();

		fireEvent.click(dropdownButton);

		expect(popover.classList.contains('show')).toBeTruthy();

		// Structure level option

		expect(options[0].checked).toBeTruthy();
	});

	it('renders the popover with view level option selected', () => {
		const {container} = render(
			<LabelFieldWrapper state={VIEW_LEVEL_STATE} />
		);

		const dropdownButton = container.querySelector(
			'.form-renderer-label-field__button'
		);
		const popover = container.querySelector(
			'.form-renderer-label-field__popover'
		);
		const options = document.querySelectorAll(
			'.form-renderer-label-field input.custom-control-input'
		);

		expect(popover.classList.contains('show')).toBeFalsy();

		fireEvent.click(dropdownButton);

		expect(popover.classList.contains('show')).toBeTruthy();

		// View level option

		expect(options[1].checked).toBeTruthy();
	});

	it('change the label after choose the popover option', () => {
		const {container} = render(
			<LabelFieldWrapper state={VIEW_LEVEL_STATE} />
		);

		const options = document.querySelectorAll(
			'.form-renderer-label-field input.custom-control-input'
		);
		const input = container.querySelector('.ddm-field-text');

		// Structure level option

		fireEvent.click(options[0]);

		expect(input.value).toBe('Username');

		// View Level option

		fireEvent.click(options[1]);

		expect(input.value).toBe('Username');
	});
});
