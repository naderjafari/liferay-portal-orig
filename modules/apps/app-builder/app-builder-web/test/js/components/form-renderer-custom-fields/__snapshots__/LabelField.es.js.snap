// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabelField renders with structure-level 1`] = `
<DocumentFragment>
  <div
    class="form-renderer-label-field d-flex justify-content-between"
  >
    <div
      class="form-group form-renderer-label-field__input"
    >
      <label
        class="ddm-label"
      >
        Label
        <span
          class="ddm-tooltip"
        >
          <svg
            class="lexicon-icon lexicon-icon-question-circle-full"
            role="presentation"
            title="Enter a descriptive field label that guides users to enter the information you want."
          >
            <use
              xlink:href="#question-circle-full"
            />
          </svg>
        </span>
      </label>
      <input
        class="form-control ddm-field-text"
        name="LabelName"
        placeholder="Enter a field label."
        type="text"
        value="Name"
      />
    </div>
    <button
      class="form-renderer-label-field__button btn btn-monospaced btn-outline-borderless btn-sm btn-outline-secondary"
      title="label-options"
      type="button"
    >
      <svg
        class="lexicon-icon lexicon-icon-ellipsis-v"
        role="presentation"
      >
        <use
          xlink:href="#ellipsis-v"
        />
      </svg>
    </button>
    <div
      class="form-renderer-label-field__popover popover clay-popover-bottom-right"
    >
      <div
        class="arrow"
      />
      <div
        class="inline-scroller"
        tabindex="-1"
      >
        <div
          class="popover-header"
        >
          label-options
        </div>
        <div
          class="popover-body"
        >
          <div
            class="mt-2"
          >
            <div
              class=""
            >
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    checked=""
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="structure-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      label-for-all-forms-using-this-field
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="view-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      label-for-only-this-form
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`LabelField renders with view-level 1`] = `
<DocumentFragment>
  <div
    class="form-renderer-label-field d-flex justify-content-between"
  >
    <div
      class="form-group form-renderer-label-field__input"
    >
      <label
        class="ddm-label"
      >
        Label
        <span
          class="ddm-tooltip"
        >
          <svg
            class="lexicon-icon lexicon-icon-question-circle-full"
            role="presentation"
            title="Enter a descriptive field label that guides users to enter the information you want."
          >
            <use
              xlink:href="#question-circle-full"
            />
          </svg>
        </span>
      </label>
      <input
        class="form-control ddm-field-text"
        name="LabelName"
        placeholder="Enter a field label."
        type="text"
        value="Username"
      />
      <div
        class="form-feedback-group"
      >
        <div
          class="form-feedback-item"
        >
          object-field-label-x
        </div>
      </div>
    </div>
    <button
      class="form-renderer-label-field__button btn btn-monospaced btn-outline-borderless btn-sm btn-outline-secondary"
      title="label-options"
      type="button"
    >
      <svg
        class="lexicon-icon lexicon-icon-ellipsis-v"
        role="presentation"
      >
        <use
          xlink:href="#ellipsis-v"
        />
      </svg>
    </button>
    <div
      class="form-renderer-label-field__popover popover clay-popover-bottom-right"
    >
      <div
        class="arrow"
      />
      <div
        class="inline-scroller"
        tabindex="-1"
      >
        <div
          class="popover-header"
        >
          label-options
        </div>
        <div
          class="popover-body"
        >
          <div
            class="mt-2"
          >
            <div
              class=""
            >
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="structure-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      label-for-all-forms-using-this-field
                    </span>
                  </span>
                </label>
              </div>
              <div
                class="custom-control custom-radio"
              >
                <label>
                  <input
                    checked=""
                    class="custom-control-input"
                    role="radio"
                    type="radio"
                    value="view-level"
                  />
                  <span
                    class="custom-control-label"
                  >
                    <span
                      class="custom-control-label-text"
                    >
                      label-for-only-this-form
                    </span>
                  </span>
                </label>
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
